### ISSUE TITLE
squid with workers 4 option

### ISSUE THREAD

**Issue by chronolz** [termux/termux-packages#5651]
**Problem description**

squid is not working with workers option set to 4

**Steps to reproduce**

I am running squid -z under proot -0 with workers option set to 4 in squid.conf

**Expected behavior**

**Additional information**

```

2020/08/05 12:03:23| Created PID file (/data/data/com.termux/files/usr/var/run/squid.pid)
2020/08/05 12:03:23 kid1| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:23 kid1|   Finished.  Wrote 0 entries.
2020/08/05 12:03:23 kid1|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:23 kid1| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:23 kid1| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.055 seconds = 0.022 user + 0.033 sys
Maximum Resident Size: 45120 KB
Page faults with physical i/o: 0
2020/08/05 12:03:23 kid2| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:23 kid2|   Finished.  Wrote 0 entries.
2020/08/05 12:03:23 kid2|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:23 kid2| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:23 kid2| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.052 seconds = 0.000 user + 0.052 sys
Maximum Resident Size: 44544 KB
Page faults with physical i/o: 0
2020/08/05 12:03:23 kid3| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:23 kid3|   Finished.  Wrote 0 entries.
2020/08/05 12:03:23 kid3|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:23 kid3| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:23 kid3| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.058 seconds = 0.029 user + 0.029 sys
Maximum Resident Size: 45136 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid4| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid4|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid4|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid4| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid4| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.077 seconds = 0.039 user + 0.038 sys
Maximum Resident Size: 45664 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid5| Set Current Directory to /data/data/com.termux/files/usr/var/cache/squid
2020/08/05 12:03:24 kid5| Creating missing swap directories
2020/08/05 12:03:24 kid5| No cache_dir stores are configured.
2020/08/05 12:03:24 kid1| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid1|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid1|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid1| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid1| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.081 seconds = 0.016 user + 0.065 sys
Maximum Resident Size: 45536 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid2| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid2|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid2|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid2| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid2| Squid Cache (Version 4.12): Terminated abnormally.
2020/08/05 12:03:24 kid3| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid3|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid3|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid3| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid3| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.093 seconds = 0.047 user + 0.046 sys
Maximum Resident Size: 45056 KB
Page faults with physical i/o: 0
CPU Usage: 0.094 seconds = 0.032 user + 0.062 sys
Maximum Resident Size: 45040 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid4| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid4|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid4|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid4| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid4| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.097 seconds = 0.065 user + 0.033 sys
Maximum Resident Size: 45232 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid3| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid3|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid3|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid3| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid3| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.061 seconds = 0.041 user + 0.020 sys
Maximum Resident Size: 45136 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid2| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid2|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid2|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid2| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid2| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.063 seconds = 0.047 user + 0.016 sys
Maximum Resident Size: 45056 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid1| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid1|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid1|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid1| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid1| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.076 seconds = 0.044 user + 0.032 sys
Maximum Resident Size: 44928 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid4| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid4|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid4|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid4| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid4| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.060 seconds = 0.030 user + 0.030 sys
Maximum Resident Size: 45616 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid2| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid2|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid2|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid2| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid2| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.043 seconds = 0.022 user + 0.022 sys
Maximum Resident Size: 45088 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid3| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid3|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid3|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid3| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid3| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.057 seconds = 0.034 user + 0.023 sys
Maximum Resident Size: 45232 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid1| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid1|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid1|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid1| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid1| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.054 seconds = 0.027 user + 0.027 sys
Maximum Resident Size: 44928 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid3| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid3|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid3|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid3| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid3| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.047 seconds = 0.023 user + 0.023 sys
Maximum Resident Size: 45184 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid1| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid1|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid1|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid1| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid1| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.047 seconds = 0.031 user + 0.016 sys
Maximum Resident Size: 44976 KB
Page faults with physical i/o: 0
2020/08/05 12:03:24 kid4| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid4|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid4|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid4| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid2| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:24 kid2|   Finished.  Wrote 0 entries.
2020/08/05 12:03:24 kid2|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:24 kid2| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:24 kid4| Squid Cache (Version 4.12): Terminated abnormally.
2020/08/05 12:03:24 kid2| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.062 seconds = 0.041 user + 0.021 sys
Maximum Resident Size: 44416 KB
Page faults with physical i/o: 0
CPU Usage: 0.065 seconds = 0.039 user + 0.026 sys
Maximum Resident Size: 44640 KB
Page faults with physical i/o: 0
2020/08/05 12:03:25 kid4| storeDirWriteCleanLogs: Starting...
2020/08/05 12:03:25 kid4|   Finished.  Wrote 0 entries.
2020/08/05 12:03:25 kid4|   Took 0.00 seconds (  0.00 entries/sec).
2020/08/05 12:03:25 kid4| FATAL: Ipc::Mem::Segment: Cannot fake shared segments in SMP config (Fake segment creation)

2020/08/05 12:03:25 kid4| Squid Cache (Version 4.12): Terminated abnormally.
CPU Usage: 0.315 seconds = 0.064 user + 0.251 sys
Maximum Resident Size: 44496 KB
Page faults with physical i/o: 0
2020/08/05 12:03:25| Removing PID file (/data/data/com.termux/files/usr/var/run/squid.pid)


```

**Comment by xeffyr** [comment]
Why you are running it under proot? And what configuration you are using?

Note that our `squid` build supports only in-memory caching, IPC can also be problematic (especially under proot).

**Comment by chronolz** [comment]
I am running under proot because of issue #5388 "Squid : Bad System Call"

I am running squid with the standard config file and the following options
workers 4
cache deny all

the workers 4 option opens parallel instances of squid for managing workload

With conventional distros squid uses using /dev/shm, see issue #3485 "Squid failing to run" that this build may be faking? That faking perhaps isn't playing nice with SMP... maybe this is an easy fix?

Thanks

**Comment by xeffyr** [comment]
> With conventional distros squid uses using /dev/shm

POSIX shared memory is not supported on Android.

I can patch it to enable support for POSIX shmem, but as Android doesn't provide /dev/shm or other world-writable TMPFS directory, memory blobs will be written to disk potentially reducing lifetime of the device's flash storage.

> maybe this is an easy fix?

Easy fix would be to root device and setup a normal Linux chroot with all required mount points and install squid here.

As I wrote previously, our squid build does not support disk caching, so option `-z` is useless. Only RAM-based caching is possible. Want disk cache, use package `polipo` instead.

Disk cache is not supported because Android lacks required libc features.

**Comment by xeffyr** [comment]
> Disk cache is not supported because Android lacks required libc features.

Ah, not because of missing libc features but due to missing support of POSIX message queues in kernel. That's definitely can't be fixed on Termux side and even root-based chroot won't help.

**Comment by chronolz** [comment]
I am ok with not using the -z flag. And I can run squid if i dont use the **workers** (squid with smp) option in squid config. My question is, how was issue  #3485 "Squid failing to run" resolved and if that same solution could extended to squid with smp so I can run squid with the **workers** option

**Comment by xeffyr** [comment]
> how was issue #3485 "Squid failing to run" resolved and if that same solution could be extended to squid with smp

This issue was about creating log directory, not about smp:
```
WARNING: Cannot write log file: /data/data/com.termux/files/usr/var/logs/cache.log
/data/data/com.termux/files/usr/var/logs/cache.log: No such file or directory
```
Solution was:
```
mkdir -p /data/data/com.termux/files/usr/var/logs
```

**Comment by xeffyr** [comment]
I can patch it and add support for using `shm_open()` functions, that will fix `Ipc::Mem::Segment`. But as I wrote previously, this will have performance penalty and affect flash storage lifetime (depending on usage), as it will place shmem segments to $PREFIX/tmp instead of /dev/shm as one is unsupported on Android.

**Comment by xeffyr** [comment]
Fixed in https://github.com/termux/termux-packages/pull/5670. It adds support for POSIX shmem and makes squid usable without proot.

Updated package will be available soon.

**Comment by chronolz** [comment]
Thank you, i appreciate it.

**Comment by chronolz** [comment]
Hi, I have tested the patch and squid now works without proot -0. Its also running in smp with workers option set to a # (I have it set to 4) although I had to create a directory to fix a socket bind error.

makedir -p ../usr/var/run/squid

https://wiki.squid-cache.org/Features/SmpScale#Feature:_SMP_Scalability
"Also may display as Cannot bind socket FD NN to 0.0.0.0: (13) Permission denied if you disabled IPv6.

Squid registers its IPC channel sockets (Unix Domain Sockets or UDS) in the localstatedir. For standard installs, this is usually /var/run/squid. If you installed Squid to a different directory, see the --localstatedir ./configure option.

Check the permissions on /var/run/squid (or whatever your localstatedir is). The localstatedir directory requires ownership by the Squid user for SMP Squid to work."
